enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,724 @@
+    // This number will be balanced out across all specified racks.
+    // Cannot be used together with `nodes` specified within any rack spec.
+    // +optional
+    Nodes *int32 `json:"nodes"`
+
+    // racks specify the racks in the datacenter.
+    // +optional
+    Racks []RackSpec `json:"racks"`
+
+    // automaticOrphanedNodeCleanup controls if automatic orphan node cleanup should be performed.
+    AutomaticOrphanedNodeCleanup bool `json:"automaticOrphanedNodeCleanup,omitempty"`
+
+    // sysctls holds the sysctl properties to be applied during initialization given as a list of key=value pairs.
+    // Example: fs.aio-max-nr=232323
+    // +optional
+    Sysctls []string `json:"sysctls,omitempty"`
+
+    // externalSeeds specifies the external seeds to propagate to ScyllaDB binary on startup as "seeds" parameter of seed-provider.
+    ExternalSeeds []string `json:"externalSeeds,omitempty"`
+
+    // minTerminationGracePeriodSeconds specifies minimum duration in seconds to wait before every drained node is
+    // terminated. This gives time to potential load balancer in front of a node to notice that node is not ready anymore
+    // and stop forwarding new requests.
+    // This applies only when node is terminated gracefully.
+    // If not provided, Operator will determine this value.
+    // EXPERIMENTAL. Do not rely on any particular behaviour controlled by this field.
+    // +optional
+    MinTerminationGracePeriodSeconds *int32 `json:"minTerminationGracePeriodSeconds,omitempty"`
+
+    // minReadySeconds is the minimum number of seconds for which a newly created ScyllaDB node should be ready
+    // for it to be considered available.
+    // When used to control load balanced traffic, this can give the load balancer in front of a node enough time to
+    // notice that the node is ready and start forwarding traffic in time. Because it all depends on timing, the order
+    // is not guaranteed and, if possible, you should use readinessGates instead.
+    // If not provided, Operator will determine this value.
+    // +optional
+    MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+
+    // readinessGates specifies custom readiness gates that will be evaluated for every ScyllaDB Pod readiness.
+    // It's projected into every ScyllaDB Pod as its readinessGate. Refer to upstream documentation to learn more
+    // about readiness gates.
+    // +optional
+    ReadinessGates []corev1.PodReadinessGate `json:"readinessGates,omitempty"`
+}
+
+type ObjectTemplateMetadata struct {
+    // labels is a custom key value map that gets merged with managed object labels.
+    // +optional
+    Labels map[string]string `json:"labels,omitempty"`
+
+    // annotations is a custom key value map that gets merged with managed object annotations.
+    // +optional
+    Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+// ScyllaDBOverrides allows to override a subset of ScyllaDB settings.
+type ScyllaDBOverrides struct {
```

Comment:
This struct kind of requires us to define the racks and partially options on ScyllaDBDatacenterSpec again and let only certain fields be adjusted per rack. Maybe even the image and others could be adjusted per rack for blue-green updates or testing / debugging.

I am wondering whether we could somehow define this once and use it with the overrides again. Like by defining a RackBase struct that would be embedded at the spec level and also be able to optionally set in for each rack. Something like "defaultRack" and then defining the array of names and topology label selectors unless you want to adjust it.

Slightly related idea that has been lingering in my head for quite a while is to create a `ScyllaDBRack` to split the API responsibilities and controllers, allow API based adjustments and possibly manage Pods directly at some point, if needed. It may be a step `#2` but it would also give us `RackTemplateSpec` to use.

Reply:
to make the naming in the above more meaningfull, how about something like
```yaml
spec:
  rackTemplate:
    scylladb:
      resources:    
     ...
  racks:
  - name: foo
  - name: bar
    rackTemplate:
      scylladb:
        resources:
```

Reply:
I had similar remarks and we discussed them with @zimnx yesterday.  In my opinion there shouldn't be a split between the template and per-rack level (in other words the overrides shouldn't be narrower) - seems like we agree on this one. To add to this comment though, you'd still need global values like the proposed `nodes` (https://github.com/scylladb/scylla-operator/pull/1994#discussion_r1662331391) which don't fall into the `rackTemplate`.
So to cover the datacenter-level fields you could extend it with something like:

```
spec:
  scylladb:
    nodes:
    enableDeveloperMode:
    ...
  rackTemplate:
    scylladb:
      resources:    
     ...
  racks:
  - name: foo
  - name: bar
    rackTemplate:
      scylladb:
        resources:
```

Reply:
I mostly agree but `nodes` shouldn't be under scylladb (the apply to manager agent / the whole Pod equally

Reply:
> the apply to manager agent / the whole Pod equally

yes, but only because its implied by the number of Scylla nodes. I'm not saying it has to be under `scylladb`, but that's not a reason not to imo

Reply:
done